addons/hosted-by-archive/jaxrs/src/main/java/org/commonjava/indy/hostedbyarc/bind/jaxrs/IndyHostedByArchiveResource.java

```diff
@@ -0,0 +1,165 @@
+import org.commonjava.indy.util.ApplicationContent;
+import org.commonjava.indy.util.ApplicationStatus;
+import org.jboss.resteasy.annotations.providers.multipart.MultipartForm;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.core.SecurityContext;
+import javax.ws.rs.core.UriInfo;
+
+import static org.commonjava.indy.bind.jaxrs.util.ResponseUtils.formatResponse;
+import static org.commonjava.indy.pkg.maven.model.MavenPackageTypeDescriptor.MAVEN_PKG_KEY;
+
+@Api( value = "Hosted by archive", description = "Create a new maven hosted store by zip file" )
+@Path( "/api/admin/stores/maven/hosted/{name}" )
+@ApplicationScoped
+public class IndyHostedByArchiveResource
+        implements IndyResources
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private HostedByArchiveManager hostedByArchiveManager;
+
+    @Inject
+    private AdminController adminController;
+
+    @Inject
+    private SecurityManager securityManager;
+
+    @Inject
+    private HostedByArchiveConfig config;
+
+    @ApiOperation( "Create a new maven hosted store by a zip file" )
+    @ApiResponses( { @ApiResponse( code = 201, response = ArtifactStore.class, message = "The store was created" ),
+                           @ApiResponse( code = 409,
+                                         message = "A store with the specified type and name already exists" ) } )
+    @ApiImplicitParams( { @ApiImplicitParam( paramType = "body", name = "body", required = true,
+                                             dataType = "org.commonjava.indy.model.core.ArtifactStore",
+                                             value = "The artifact store definition JSON" ) } )
+    @Path( "/compressed-content" )
+    @POST
+    @Consumes( MediaType.MULTIPART_FORM_DATA )
+    @Produces( ApplicationContent.application_json )
+    public Response postCreateHostedByZip( final @PathParam( "name" ) String name, final @Context UriInfo uriInfo,
+                                           final @QueryParam( "ignoreBeforeRoot" ) String ignoreBeforeRoot,
+                                           final @MultipartForm FileUploadForm form,
+                                           final @Context HttpServletRequest request,
+                                           final @Context SecurityContext securityContext )
+    {
+        return createHostedByZip( name, uriInfo, ignoreBeforeRoot, form, request, securityContext );
```

Comment:
I wonder if we can probe these for Zip compat here? I did some looking around using hexedit, and zip files all seem to have the same header: `0x50 0x4B 0x03 0x04` that we might be able to read to determine whether this is likely to be a zip file we can open.

I'd also like us to be able to consume the HttpServletRequest InputStream from an endpoint marked with @Consumes("application/zip") via POST / PUT, since this will be a little more machine-friendly than constructing a form post (I think).